plugins/connection/redirection_utility.c

```diff
@@ -0,0 +1,555 @@
+   modify it under the terms of the GNU Library General Public
+   License as published by the Free Software Foundation; either
+   version 2 of the License, or (at your option) any later version.
+
+   This library is distributed in the hope that it will be useful,
+   but WITHOUT ANY WARRANTY; without even the implied warranty of
+   MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
+   Library General Public License for more details.
+
+   You should have received a copy of the GNU Library General Public
+   License along with this library; if not see <http://www.gnu.org/licenses>
+   or write to the Free Software Foundation, Inc.,
+   51 Franklin St., Fifth Floor, Boston, MA 02110, USA
+*************************************************************************************/
+
+#include <ma_global.h>
+#include <ma_sys.h>
+#include <mysql.h>
+#include <errmsg.h>
+#include <mysql/client_plugin.h>
+#include <string.h>
+#include <ma_common.h>
+#include <mariadb_ctype.h>
+#include "ma_server_error.h"
+#include <ma_pthread.h>
+#include "redirection_utility.h"
+
+// Note that key, host and user and malloced together with Redirection_Entry
+struct Redirection_Entry
+{
+	char* key;
+	char* host;
+	char* user;
+	unsigned int port;
+	struct Redirection_Entry* next;
+	struct Redirection_Entry* prev;
+};
+
+// Note that host and user are malloced together with Redirection_info
+struct Redirection_Info
+{
+	char* host;
+	char* user;
+	unsigned int port;
+};
+
+#define MAX_CACHED_REDIRECTION_ENTRYS 4
+#define MAX_REDIRECTION_KEY_LENGTH 512
+#define MAX_REDIRECTION_INFO_LENGTH 512
+static unsigned int redirection_cache_num = 0;
+static struct Redirection_Entry redirection_cache_root = { 0 };
+static struct Redirection_Entry* redirection_cache_rear = NULL;
+
+static pthread_mutex_t redirect_lock;
+
+#define redirect_mutex redirect_lock
+
+#define mutex_lock(x) pthread_mutex_lock(&x);
+
+#define mutex_unlock(x) pthread_mutex_unlock(&x)
+
+#define mutex_destroy(x) pthread_mutex_destroy(&x)
+
+int init_redirection_cache()
+{
+	pthread_mutex_init(&redirect_mutex, NULL);
+
+	redirection_cache_root.next = NULL;
+	redirection_cache_rear = &redirection_cache_root;
+	redirection_cache_num = 0;
+
+	return 0;
+}
+
+void add_redirection_entry(char* key, char* host, char* user, unsigned int port);
+
+void delete_redirection_entry(char* key);
+
+my_bool get_redirection_info(char* key, struct Redirection_Info** info, my_bool copy_on_found);
+
+struct Redirection_Info* parse_redirection_info(MYSQL* mysql);
+
+MYSQL* check_redirect(MYSQL* mysql, const char* host,
+	const char* user, const char* passwd,
+	const char* db, uint port,
+	const char* unix_socket,
+	ulong client_flag)
+{
+	if (mysql->options.redirection_mode != REDIRECTION_OFF)
+	{
+		const char redirect_key[MAX_REDIRECTION_KEY_LENGTH] = { 0 };
+		struct Redirection_Info* info = NULL;
+
+		if (!port)
+			port = MARIADB_PORT;
+
+		sprintf(redirect_key, "%s_%s_%u", host, user, port);
```

Comment:
Better to use snprintf: If max_length for user (currently 80 * 3) or host (60 * 3) will change (or charset will change from utf8 to utf8mb4) you might get a buffer overflow.